File: 算法/代码随想录算法PDF/00调试源代码/714.买卖股票的最佳时机含手续费.py
#
# @lc app=leetcode.cn id=714 lang=python3
#
# [714] 买卖股票的最佳时机含手续费
#

# @lc code=start
class Solution:
    def maxProfit(self, prices: List[int], fee: int) -> int:
        res = 0
        minPrice = prices[0]
        for i in range(1, len(prices)):
            if prices[i] < minPrice:
                minPrice = prices[i]
            if prices[i] > minPrice + fee:
                res += prices[i] - minPrice - fee
                minPrice = prices[i] - fee  # 重要！
        return res
# @lc code=end

File: 算法/代码随想录算法PDF/00调试源代码/714.买卖股票的最佳时机含手续费.cpp
/*
 * @lc app=leetcode.cn id=714 lang=cpp
 *
 * [714] 买卖股票的最佳时机含手续费
 */

// @lc code=start
class Solution {
public:
    int maxProfit(vector<int>& prices, int fee) {
        int res = 0;
        int minPrice = prices[0];
        for (int i = 1; i < prices.size(); i++) {
            if (prices[i] < minPrice) minPrice = prices[i];
            if (prices[i] > minPrice + fee) {
                res += prices[i] - minPrice - fee;
                minPrice = prices[i] - fee;  // 重要！用于在有更高价格时后悔
            }
        }
        return res;
    }
};
// @lc code=end

